### ISSUE TITLE
[Flang] Incorrect diagnose when referencing an implicitly typed procedure pointer within a module

### ISSUE THREAD

**Issue by DanielCChen** [llvm/llvm-project#83622]
Consider the following code:
```
  MODULE M
  IMPLICIT INTEGER(P)

  PROCEDURE(), POINTER :: ProcPtr

  CONTAINS

  FUNCTION F()
  INTEGER F
  END FUNCTION

  SUBROUTINE S()
    ProcPtr => F
    IF ( ProcPtr() .NE. 1 ) ERROR STOP 11
  END SUBROUTINE

  END
  ```
  
  Flang currently issues an error:
  ```
./t.f:14:10: error: Cannot call subroutine 'procptr' like a function
      IF ( ProcPtr() .NE. 1 ) ERROR STOP 11
           ^^^^^^^
./t.f:14:10: Declaration of 'procptr'
      IF ( ProcPtr() .NE. 1 ) ERROR STOP 11
           ^^^^^^^
```

The code seems confirming to me. `ProcPtr` is implicitly typed to integer and pointer assignment to an integer function and referenced as a function.
If I put these code in a program unit instead of a module. It seems compiles pass the error.

**Comment by klausler** [comment]
https://github.com/llvm/llvm-project/pull/83897

**Comment by DanielCChen** [comment]
I have verified that the reproducer and the original test case are fixed by PR #83897.
